File: Python3/_1959_MinimumTotalSpaceWastedWithKResizingOperations.py
from typing import List


class Solution:
    def minSpaceWastedKResizing(self, nums: List[int], k: int) -> int:
        memory = [ [-1] * 200 for i in range(200) ]
        return self.dp(nums,memory,0,k)

    def dp(self, nums: List[int], memory: List[int], i: int, k: int)-> int:
        if i == len(nums):
            return 0
        if k < 0 :
            return 200000000
        if memory[i][k] >= 0:
            return memory[i][k]
        
        sum = 0
        maxVal = nums[i]
        ans = 200000000
        for j in range(i,len(nums)):
            sum += nums[j]
            maxVal = max(maxVal,nums[j])
            waste = maxVal*(j-i+1) - sum
            ans = min(ans,self.dp(nums,memory,j+1,k-1)+waste)
        memory[i][k] = ans
        return ans

File: Csharp/_1959_MinimumTotalSpaceWastedWithKResizingOperations.cs
using System;
using System.Collections.Generic;

namespace Csharp
{
    public class _1959_MinimumTotalSpaceWastedWithKResizingOperations {
        public int MinSpaceWastedKResizing(int[] nums, int k) {
            int[][] memeories = new int[200][];
            for(int i = 0; i < memeories.Length; i++){
                memeories[i] = new int[200];
                for(int j = 0; j < memeories[i].Length; j++)
                    memeories[i][j] = -1;
            }
            return dp(nums,memeories,0,k);
        }

        private int dp(int[] nums, int[][] memories, int i, int k){
            if(i == nums.Length ) return 0;
            if(k < 0) return 200000000;
            if(memories[i][k] >= 0) return memories[i][k]; 

            int ans = 200000000;
            int max = nums[i];
            int sum = 0;
            for(int j = i; j<nums.Length; j++){
                max = Math.Max(max,nums[j]);
                sum += nums[j];
                int waste = max*(j-i+1) - sum;
                ans = Math.Min(ans,waste + dp(nums,memories,j+1,k-1));
            } 

            memories[i][k] = ans;
            return ans;
        }
    }
}